File: src/transaction.js
import crypto from "node:crypto";
import { makeString, setMerchantCity, setMerchantName } from "./convert.js";
import QRCode from "qrcode";
import { parseQRIS } from "./tlv.js";

const DEFAULT_TTL_MS = 15 * 60 * 1000;

export class TransactionStore {
  constructor({ ttlMs = DEFAULT_TTL_MS } = {}) {
    this.ttlMs = ttlMs;
    this.map = new Map();
  }

  create({ qris, nominal, customerName, description, fee = "0", taxtype = "p", city, name, refPrefix = "RAOL" }) {
    const ref = refPrefix + Date.now().toString(36).toUpperCase() + crypto.randomBytes(2).toString("hex").toUpperCase();
    let base = qris;
    if (city) base = setMerchantCity(base, city);
    if (name) base = setMerchantName(base, name);

    const qrisString = makeString(base, { nominal: String(nominal), fee: String(fee), taxtype });
    const expiresAt = Date.now() + this.ttlMs;

    const tx = {
      ref,
      nominal: Number(nominal),
      customerName: customerName || null,
      description: description || null,
      fee: Number(fee),
      taxtype,
      qrisString,
      status: "pending",
      createdAt: new Date().toISOString(),
      expiresAt: new Date(expiresAt).toISOString(),
      paidAt: null,
    };
    this.map.set(ref, tx);
    return tx;
  }

  get(ref) {
    const tx = this.map.get(ref);
    if (!tx) return null;
    if (tx.status === "pending" && Date.now() > new Date(tx.expiresAt).getTime()) {
      tx.status = "expired";
    }
    return tx;
  }

  confirm(ref, { method = "manual" } = {}) {
    const tx = this.map.get(ref);
    if (!tx) return null;
    if (tx.status === "expired") throw new Error("transaksi sudah expired");
    tx.status = "paid";
    tx.paidAt = new Date().toISOString();
    tx.paidMethod = method;
    return tx;
  }

  list({ status } = {}) {
    const all = [...this.map.values()];
    return status ? all.filter((t) => t.status === status) : all;
  }

  purgeExpired() {
    const now = Date.now();
    for (const [ref, tx] of this.map) {
      if (tx.status === "pending" && now > new Date(tx.expiresAt).getTime()) {
        tx.status = "expired";
      }
    }
  }
}

export async function receiptQR(tx) {
  const qrImage = await QRCode.toDataURL(tx.qrisString, { margin: 2, width: 420, errorCorrectionLevel: "M" });
  return { ...tx, qrImage, parsed: parseQRIS(tx.qrisString) };
}

File: mobile/lib/qris/transaction.dart
import 'dart:math';
import 'convert.dart';
import 'tlv.dart';

class Transaction {
  Transaction({
    required this.ref,
    required this.nominal,
    this.customerName,
    this.description,
    this.fee = 0,
    this.taxtype = "p",
    required this.qrisString,
    required this.status,
    required this.createdAt,
    required this.expiresAt,
    this.paidAt,
    this.paidMethod,
  });

  final String ref;
  final int nominal;
  final String? customerName;
  final String? description;
  final int fee;
  final String taxtype;
  final String qrisString;
  String status;
  final String createdAt;
  final String expiresAt;
  String? paidAt;
  String? paidMethod;
}

class TransactionStore {
  TransactionStore({this.ttlMs = 15 * 60 * 1000});
  final int ttlMs;
  final Map<String, Transaction> _map = {};

  Transaction create({
    required String qris,
    required String nominal,
    String? customerName,
    String? description,
    String fee = "0",
    String taxtype = "p",
    String? city,
    String? name,
    String refPrefix = "RAOL",
  }) {
    String base = qris;
    if (city != null) base = setMerchantCity(base, city);
    if (name != null) base = setMerchantName(base, name);
    final qrisString = makeString(base, nominal: nominal, fee: fee, taxtype: taxtype);
    final ref = refPrefix +
        DateTime.now().millisecondsSinceEpoch.toRadixString(36).toUpperCase() +
        Random().nextInt(65535).toRadixString(16).toUpperCase().padLeft(4, '0');
    final now = DateTime.now();
    final tx = Transaction(
      ref: ref,
      nominal: int.parse(nominal),
      customerName: customerName,
      description: description,
      fee: int.tryParse(fee) ?? 0,
      taxtype: taxtype,
      qrisString: qrisString,
      status: "pending",
      createdAt: now.toIso8601String(),
      expiresAt: now.add(Duration(milliseconds: ttlMs)).toIso8601String(),
    );
    _map[ref] = tx;
    return tx;
  }

  Transaction? get(String ref) {
    final tx = _map[ref];
    if (tx == null) return null;
    if (tx.status == "pending" && DateTime.now().isAfter(DateTime.parse(tx.expiresAt))) {
      tx.status = "expired";
    }
    return tx;
  }

  Transaction? confirm(String ref, {String method = "manual"}) {
    final tx = _map[ref];
    if (tx == null) return null;
    if (tx.status == "expired") throw ArgumentError("transaksi sudah expired");
    tx.status = "paid";
    tx.paidAt = DateTime.now().toIso8601String();
    tx.paidMethod = method;
    return tx;
  }

  List<Transaction> list({String? status}) {
    final all = _map.values.toList();
    return status == null ? all : all.where((t) => t.status == status).toList();
  }
}
